vpc/service/allocate.go

```diff
@@ -0,0 +1,379 @@
+		"region":   rq.AddressAllocation.AddressLocation.Region,
+		"az":       rq.AddressAllocation.AddressLocation.AvailabilityZone,
+		"subnetId": rq.AddressAllocation.AddressLocation.SubnetId,
+	})
+	logger.G(ctx).Info("Performing Address allocation")
+
+	subnet, err := session.GetSubnetByID(ctx, rq.AddressAllocation.AddressLocation.SubnetId, ec2wrapper.UseCache)
+	if err != nil {
+		logger.G(ctx).WithError(err).Error("Could not get subnet")
+		return nil, err
+	}
+
+	if aws.StringValue(subnet.AvailabilityZone) != rq.AddressAllocation.AddressLocation.GetAvailabilityZone() {
+		return nil, status.Errorf(codes.InvalidArgument, "Subnet %q is in availability zone %q, while specified address location is in availability zone %q", aws.StringValue(subnet.SubnetId), aws.StringValue(subnet.AvailabilityZone), rq.AddressAllocation.AddressLocation.AvailabilityZone)
+	}
+
+	vpcService.dummyInterfaceLock.Lock()
+	defer vpcService.dummyInterfaceLock.Unlock()
+
+	dummyNetworkInterfaceSession, dummyNetworkInterface, err := vpcService.getDummyInterface(ctx, session, subnet)
+	if err != nil {
+		logger.G(ctx).WithError(err).Error("Could not get dummy interface")
+		return nil, err
+	}
+
+	awsSession, err := dummyNetworkInterfaceSession.Session(ctx)
+	if err != nil {
+		return nil, ec2wrapper.HandleEC2Error(err, span)
+	}
+
+	logger.G(ctx).WithField("networkInterfaceId", aws.StringValue(dummyNetworkInterface.NetworkInterfaceId)).Info("Got network interface")
+
+	tx, err := vpcService.conn.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not start database transaction")
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	rows, err := tx.QueryContext(ctx, "SELECT id FROM ip_addresses WHERE id = $1 FOR UPDATE LIMIT 1", allocationUUID)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not query database for existing records")
+	}
+	defer rows.Close()
+	if rows.Next() {
+		return nil, status.Errorf(codes.AlreadyExists, "UUID %s is already allocated", allocationUUID)
+	}
+
+	assignPrivateIPAddressesOutput, err := dummyNetworkInterfaceSession.AssignPrivateIPAddresses(ctx, ec2.AssignPrivateIpAddressesInput{
+		SecondaryPrivateIpAddressCount: aws.Int64(1),
+	})
+	if err != nil {
+		return nil, ec2wrapper.HandleEC2Error(err, span)
+	}
+
+	if count := len(assignPrivateIPAddressesOutput.AssignedPrivateIpAddresses); count != 1 {
+		return nil, status.Errorf(codes.Internal, "Instead of receiving 1 IP upon allocation from AWS, received: %d", count)
+	}
+
+	ip := net.ParseIP(aws.StringValue(assignPrivateIPAddressesOutput.AssignedPrivateIpAddresses[0].PrivateIpAddress))
+	az := aws.StringValue(dummyNetworkInterface.AvailabilityZone)
+
+	ec2session := ec2.New(awsSession)
+	_, err = ec2session.CreateTagsWithContext(ctx, &ec2.CreateTagsInput{
+		DryRun:    nil,
+		Resources: []*string{dummyNetworkInterface.NetworkInterfaceId},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String(fmt.Sprintf("vpc-service:%s", allocationUUID)),
```

Comment:
Don't know if there's a limit on the length of keys, but it might be worth calling this `titus-vpc-service:%s`, so that people poking around AWS know who to ask about these ENIs.